src/main/core/controller.c

```diff
@@ -311,38 +332,87 @@ static void _controller_registerHostCallback(const char* name, const ConfigOptio
         params->interfaceBufSize = config_getInterfaceBuffer(config);
         params->qdisc = config_getInterfaceQdisc(config);
 
+        /* bandwidth values come from the host options and graph options */
+        bool foundBwDown = false;
+        bool foundBwUp = false;
+
+        foundBwDown |= networkgraph_nodeBandwidthDownBits(
+            controller->graph, graphNode, &params->requestedBwDownBits);
+        foundBwDown |= hostoptions_getBandwidthDown(host, &params->requestedBwDownBits);
+
+        foundBwUp |= networkgraph_nodeBandwidthUpBits(
+            controller->graph, graphNode, &params->requestedBwUpBits);
+        foundBwUp |= hostoptions_getBandwidthUp(host, &params->requestedBwUpBits);
+
+        if (!foundBwDown) {
+            error("No downstream bandwidth provided for host %s", params->hostname);
+            return -1;
+        }
+
+        if (!foundBwUp) {
+            error("No upstream bandwidth provided for host %s", params->hostname);
+            return -1;
+        }
+
+        if (params->requestedBwDownBits == 0 || params->requestedBwUpBits == 0) {
+            error("Bandwidth for host %s must be non-zero", params->hostname);
+            return -1;
+        }
+
+        /* add the host */
         manager_addNewVirtualHost(controller->manager, params);
 
+        /* now handle each virtual process the host will run */
         ProcessCallbackArgs processArgs;
         processArgs.controller = controller;
         processArgs.hostname = hostnameBuffer->str;
-
-        /* now handle each virtual process the host will run */
-        hostoptions_iterProcesses(host, _controller_registerProcessCallback, (void*)&processArgs);
+        if (hostoptions_iterProcesses(
+                host, _controller_registerProcessCallback, (void*)&processArgs)) {
+            return -1;
+        }
 
         /* cleanup for next pass through the loop */
         g_string_free(hostnameBuffer, TRUE);
-
         hostoptions_freeString(params->pcapDir);
-        hostoptions_freeString(params->ipHint);
-        hostoptions_freeString(params->countrycodeHint);
-        hostoptions_freeString(params->citycodeHint);
-
         g_free(params);
     }
+
+    // no error
+    return 0;
 }
 
-static void _controller_registerHosts(Controller* controller) {
+static int _controller_registerHosts(Controller* controller) {
     MAGIC_ASSERT(controller);
-    config_iterHosts(controller->config, _controller_registerHostCallback, (void*)controller);
+
+    RegisterHostCallbackOptions options;
+
+    // register hosts that have a specific IP address
+    options = (RegisterHostCallbackOptions){
+        .controller = controller,
+        .registerIfAddressSpecified = true,
+    };
+    if (config_iterHosts(controller->config, _controller_registerHostCallback, (void*)&options)) {
+        return -1;
```

Comment:
I don't think there's any additional information we can log here.

Reply:
If nothing else, that this is the call site of the `config_iterHosts` call that failed is useful additional information, though maybe it can be found pretty easily if there are few call sites to `config_iterHosts`. It also ensures an error is logged in case `config_iterHosts` didn't already log an error - maybe it does now, but that's a non-local assumption that could change later without this code getting updated.

Up to you, but IMO it pays to err on the side of extra verbosity when handling a fatal error.

Reply:
Added an error message here.